File: app/src/main/java/com/metalshard/hyperion/data/ScheduleRepository.kt
package com.metalshard.hyperion.data

import com.metalshard.hyperion.model.ScheduleEvent
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.net.URL

class ScheduleRepository {
    private val url = "https://pbsv.themetalshard.space/api/"
    private val gson = Gson()

    suspend fun fetchSchedule(): Map<String, List<ScheduleEvent>> = withContext(Dispatchers.IO) {
        try {
            val jsonText = URL(url).readText()

            val listType = object : TypeToken<List<Map<String, Any>>>() {}.type
            val rawList: List<Map<String, Any>> = gson.fromJson(jsonText, listType)

            val finalMap = mutableMapOf<String, List<ScheduleEvent>>()

            rawList.forEach { groupMap ->
                val groupId = groupMap["id"] as? String ?: "unknown"

                val eventsInGroup = mutableListOf<ScheduleEvent>()

                groupMap.forEach { (key, value) ->
                    if (key != "id") {
                        try {
                            val eventJson = gson.toJson(value)
                            val event = gson.fromJson(eventJson, ScheduleEvent::class.java)
                            eventsInGroup.add(event)
                        } catch (e: Exception) {
                        }
                    }
                }
                finalMap[groupId] = eventsInGroup.sortedBy { it.time }
            }
            finalMap
        } catch (e: Exception) {
            e.printStackTrace()
            emptyMap()
        }
    }
}

File: PB Schedule Viewer/ScheduleRepository.swift
import Foundation

class ScheduleRepository {
    private let url = URL(string: "https://pbsv.themetalshard.space/api/")!

    func fetchSchedule() async throws -> [String: [ScheduleEvent]] {
        let (data, _) = try await URLSession.shared.data(from: url)
        
        guard let jsonArray = try JSONSerialization.jsonObject(with: data) as? [[String: Any]] else {
            return [:]
        }
        
        var finalMap: [String: [ScheduleEvent]] = [:]
        let decoder = JSONDecoder()

        for groupMap in jsonArray {
            guard let groupId = groupMap["id"] as? String else { continue }
            
            var eventsInGroup: [ScheduleEvent] = []
            
            for (key, value) in groupMap where key != "id" {
                if let eventDict = value as? [String: Any] {
                    do {
                        let eventData = try JSONSerialization.data(withJSONObject: eventDict)
                        let event = try decoder.decode(ScheduleEvent.self, from: eventData)
                        eventsInGroup.append(event)
                    } catch {
                        print("Failed to decode event \(key): \(error)")
                    }
                }
            }
            
            finalMap[groupId] = eventsInGroup.sorted(by: { $0.time < $1.time })
        }
        
        return finalMap
    }
}
